lib/source-destination/balena-s3-source.ts

```diff
@@ -27,42 +28,111 @@ import { ZipSource } from './zip';
-	private rawSource: Http;
-	private zipSource: ZipSource;
-	private ready: Promise<void>;
-	private names: Name[] = ['balena', 'resin'];
-	public name: Name; // images can be named balena.img or resin.img
+export interface AwsCredentials {
+	accessKeyId: string;
+	secretAccessKey: string;
+}
 
-	constructor(
-		readonly bucket: string,
-		readonly deviceType: string,
-		readonly version: string,
-		readonly host = 's3.amazonaws.com',
-	) {
-		// example:
-		// bucket: resin-staging-img or resin-production-img-cloudformation
-		// deviceType: raspberry-pi
-		// version: 2.9.6+rev1.prod
+export interface BalenaS3SourceOptions {
+	host?: string; // https://s3.amazonaws.com
+	bucket?: string; // resin-staging-img or resin-production-img-cloudformation
+	prefix?: string; // images or preloaded-images or esr-images
+	deviceType: string; // raspberry-pi
+	buildId: string; // 2.9.6+rev1.prod
+	release?: string; // 1344795
+	awsCredentials?: AwsCredentials;
+}
+
+export abstract class BalenaS3SourceBase extends SourceDestination {
+	public readonly host: string;
+	public readonly bucket: string;
+	public readonly prefix: string;
+	public readonly deviceType: string;
+	public readonly buildId: string;
+	public readonly release?: string; // Only used for preloaded images
+	protected axiosInstance: AxiosInstance;
+	private static filesMissingFromPreloadedImages = [
+		'VERSION',
+		'VERSION_HOSTOS',
+		'device-type.json',
+	];
+
+	constructor({
+		host = 'https://s3.amazonaws.com',
+		bucket = 'resin-production-img-cloudformation',
+		prefix = 'images',
+		deviceType,
+		buildId,
+		release,
+		awsCredentials,
+	}: BalenaS3SourceOptions) {
 		super();
-		this.ready = this.prepare();
+		this.host = host;
+		this.bucket = bucket;
+		this.prefix = prefix;
+		this.deviceType = deviceType;
+		this.buildId = buildId;
+		this.release = release;
+		this.axiosInstance = axios.create();
+		if (awsCredentials !== undefined) {
+			this.axiosInstance.interceptors.request.use(
+				aws4Interceptor({ service: 's3' }, awsCredentials),
+			);
+		}
 	}
 
-	private async prepare(): Promise<void> {
-		this.name = await this.getName();
-		this.rawSource = new Http({ url: this.getUrl(`image/${this.name}.img`) });
-		this.zipSource = new ZipSource(
-			new Http({
-				url: this.getUrl(`image/${this.name}.img.zip`),
-				avoidRandomAccess: true,
-			}),
-		);
+	public async canCreateReadStream(): Promise<boolean> {
+		return true;
+	}
+
+	private isESR() {
+		return /^\d{4}\.\d{2}\.\d+\.(dev|prod)$/.test(this.buildId);
+	}
+
+	protected getUrl(path: string): string {
+		let prefix = this.prefix;
+		let release = this.release;
+		// Preloaded images have no VERSION, VERSION_HOSTOS or device-type.json file, we need to get it from the images or esr-images folder
+		if (
+			release !== undefined &&
+			BalenaS3SourceBase.filesMissingFromPreloadedImages.includes(path)
+		) {
+			release = undefined;
+			prefix = this.isESR() ? 'esr-images' : 'images';
```

Comment:
Can you please move these two constants to proper module-level constants?

Reply:
Added, but these values aren't used anywhere else.